R/utils_data_handling.R

```diff
@@ -1,129 +1,132 @@
+  }
 
-#' @title Pivot Range Format Forecasts From Wide to Long Format
+  return(combined[])
+}
+
+#' @title Change Data from a Sample Based Format to a Quantile Format
 #'
 #' @description
-#' Given a data.frame that follows the structure shown in
-#' [range_example_data_wide()], the function outputs the same
-#' data in a long format as (as shown in
-#' [range_example_data_long()]). This can be useful e.g. for
-#' plotting.
 #'
-#' @param data a data.frame following the specifications from
-#' [eval_forecasts()]) for quantile forecasts. For an example, see
-#' [range_example_data_wide()])
-#' @return a data.frame in long format
-#' @importFrom data.table melt
+#' Transform data from a format that is based on predictive samples to a format
+#' based on plain quantiles.
+#'
+#'
+#' @param data a data.frame with samples
+#' @param quantiles a numeric vector of quantiles to extract
+#' @param ... Additional arguments that can be passed down to [quantile()].
+#' In contrast to the default behaviour of [quantile()], this function by
+#' default has `na.rm = TRUE`
+#' @return a data.frame in a long interval range format
+#' @importFrom data.table as.data.table
+#' @importFrom stats quantile
+#' @importFrom methods hasArg
+#' @keywords data-handling
 #' @export
 #' @examples
-#' wide <- scoringutils::range_example_data_wide
-#' long <- scoringutils::range_wide_to_long(wide)
-#'
-
-range_wide_to_long <- function(data) {
-
+#' sample_to_quantile(example_integer)
+sample_to_quantile <- function(data,
+                               quantiles = c(0.05, 0.25, 0.5, 0.75, 0.95),
+                               ...) {
   data <- data.table::as.data.table(data)
-  colnames <- colnames(data)
-
-  # semi-wide format where only lower and upper are given independently
-  if (all(c("lower", "upper") %in% colnames)) {
 
-    id_vars <- colnames[!(colnames %in% c("lower", "upper"))]
-
-    # need to remove quantile column if present
-    if ("quantile" %in% colnames) {
-      data[, "quantile" := NULL]
-    }
+  reserved_columns <- c("prediction", "sample")
+  by <- setdiff(colnames(data), reserved_columns)
 
-    data <- data.table::melt(data,
-                             id.vars = id_vars,
-                             measure.vars = c("lower", "upper"),
-                             variable.name = "boundary",
-                             value.name = "prediction")
-  } else {
-    # alternative is super-wide format where every range has its own column
-    ranges <- colnames[grepl("lower", colnames) | grepl("upper", colnames)]
-
-    id_vars <- colnames[!(colnames %in% ranges)]
-
-    data <- data.table::melt(data,
-                             id.vars = id_vars,
-                             measure.vars = ranges,
-                             variable.name = "range",
-                             value.name = "prediction")
-    data[, boundary := gsub("_.*", "", range)]
-    data[, range := as.numeric(gsub("^.*?_","", range))]
+  if(!hasArg("na.rm")) {
```

Comment:
I don't think this approach will work:

``` r
f <- function(x, ...) {
  
  if (!methods::hasArg(na.rm))
    na.rm = TRUE
  
  quantile(x, ..., na.rm = na.rm)
  
}

f(1:10)
#>    0%   25%   50%   75%  100% 
#>  1.00  3.25  5.50  7.75 10.00

f(c(1:10, NA))
#>    0%   25%   50%   75%  100% 
#>  1.00  3.25  5.50  7.75 10.00

f(c(1:10, NA), na.rm = FALSE)
#> Error in quantile.default(x, ..., na.rm = na.rm): formal argument "na.rm" matched by multiple actual arguments
```

<sup>Created on 2022-03-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

The common pattern to pass default values to arguments in `...` is the following:

``` r
f <- function(x, ...) {
  
  xargs <- list(...)
  
  if (is.null(xargs$na.rm))
    xargs$na.rm <- TRUE
  
  xargs$x <- x
  
  do.call(quantile, xargs)
  
}

f(1:10)
#>    0%   25%   50%   75%  100% 
#>  1.00  3.25  5.50  7.75 10.00

f(c(1:10, NA))
#>    0%   25%   50%   75%  100% 
#>  1.00  3.25  5.50  7.75 10.00

f(c(1:10, NA), na.rm = FALSE)
#> Error in quantile.default(na.rm = FALSE, x = c(1L, 2L, 3L, 4L, 5L, 6L, : missing values and NaN's not allowed if 'na.rm' is FALSE
```

<sup>Created on 2022-03-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

It can be somewhat simplified with rlang's infix operator `%||%` (no need to explicitly depends on rlang. The code is super simple and you can copy it). But it's a matter of personal preference.

``` r
library(rlang)

f <- function(x, ...) {
  
  xargs <- list(...)
  
  xargs$na.rm <- xargs$na.rm %||% TRUE

  xargs$x <- x
  
  do.call(quantile, xargs)
  
}

f(1:10)
#>    0%   25%   50%   75%  100% 
#>  1.00  3.25  5.50  7.75 10.00

f(c(1:10, NA))
#>    0%   25%   50%   75%  100% 
#>  1.00  3.25  5.50  7.75 10.00

f(c(1:10, NA), na.rm = FALSE)
#> Error in quantile.default(na.rm = FALSE, x = c(1L, 2L, 3L, 4L, 5L, 6L, : missing values and NaN's not allowed if 'na.rm' is FALSE
```

<sup>Created on 2022-03-11 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Reply:
I had reverted it for now, but we may want to open a separate issue to implement that